core/src/main/java/org/elasticsearch/plugins/PluginsService.java

```diff
@@ -296,6 +297,13 @@ public Settings updatedSettings() {
         return services;
     }
 
+    /** Returns functions from all plugins to upgrade metadata on node startup */
+    public Collection<Function<MetaData, MetaData>> getMetaDataUpgrader() {
```

Comment:
Do this inside Node instead, no need for a public method. Just call it to collect the functions, and get them into GatewayMetaState via injection for now.